trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TransferBetweenMainAndWorker.java

```diff
@@ -0,0 +1,547 @@
+			default -> new AssertionError("Unexpected letter type: " + letter.getClass());
+			}
+			final NestedWord<LETTER> singleWord = new NestedWord<>(transferEdge(letter), nestingOperation);
+			currentWord = currentWord.concatenate(singleWord);
+
+		}
+
+		final NestedRun<LETTER, ?> workerCounterexample =
+				new NestedRun<>(currentWord, oldCounterexample.getStateSequence());
+		return workerCounterexample;
+	}
+
+	// Caches every transferred Edge bidirectional
+	private LETTER transferEdge(final LETTER letter) {
+		LETTER transferredLetter = null;
+		if (mEdgeCache.containsKey(letter)) {
+			transferredLetter = mEdgeCache.get(letter);
+		} else {
+			switch (letter) {
+			case final Call call -> transferredLetter = (LETTER) getTransferCall(call);
+			case final Return re -> transferredLetter = (LETTER) getTransferReturn(re);
+			case final StatementSequence stmt -> transferredLetter = (LETTER) getTransferStmtSequence(stmt);
+			default -> new AssertionError("Unexpected letter type: " + letter.getClass());
+			}
+		}
+		assert transferredLetter != null;
+		mEdgeCache.put(letter, transferredLetter);
+		mEdgeCache.put(transferredLetter, letter);
+		return transferredLetter;
+	}
+
+	private Call getTransferCall(final Call call) {
+		final Call newCall = new Call(call.getSerialNumber(), (BoogieIcfgLocation) call.getSource(),
+				(BoogieIcfgLocation) call.getTarget(), call.getCallStatement(), mLogger);
+		newCall.setTransitionFormula(transferTransFormulaWithMode(call.getTransformula()));
+		newCall.setPayload(call.getPayload()); // Payload is need for for example Overaprixmation Annotations
+		return newCall;
+	}
+
+	private StatementSequence getTransferStmtSequence(final StatementSequence stmt) {
+		final StatementSequence newStmt =
+				new StatementSequence(stmt.getSerialNumber(), (BoogieIcfgLocation) stmt.getSource(),
+						(BoogieIcfgLocation) stmt.getTarget(), stmt.getStatements(), mLogger);
+		newStmt.setTransitionFormula(transferTransFormulaWithMode(stmt.getTransformula()));
+		newStmt.setPayload(stmt.getPayload());
+		return newStmt;
+
+	}
+
+	private Return getTransferReturn(final Return re) {
+		final Return newReturn = new Return(re.getSerialNumber(), (BoogieIcfgLocation) re.getSource(),
+				(BoogieIcfgLocation) re.getTarget(), getTransferCall(re.getCorrespondingCall()), mLogger);
+		newReturn.setTransitionFormula(transferTransFormulaWithMode(re.getTransformula()));
+		newReturn.setPayload(re.getPayload());
+		return newReturn;
+
+	}
+
+	private UnmodifiableTransFormula transferTransFormulaWithMode(final UnmodifiableTransFormula inTF) {
+		switch (mMode) {
+		case MAIN2WORKER: {
+			return transferTransformula(inTF, mMain2worker, mWorkerScript);
+		}
+		case WORKER2MAIN: {
+			return transferTransformula(inTF, mWorker2main, mMainScript);
+		}
+		default: {
+			throw new AssertionError("Unexpected transferrer mode: " + mMode);
+		}
+		}
+	}
+
+	private UnmodifiableTransFormula transferTransformula(final UnmodifiableTransFormula inTF,
```

Comment:
We already have methods `TransferrerWithVariableCache::transferTransFormula` resp. `TransFormulaBuilder::transferTransformula`. Is your implementation doing something fundamentally different, or could you reuse those?